<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>scythe2.helpers.ensembl_ortho_mysql &mdash; Scythe 0.1 documentation</title>
    
    <link rel="stylesheet" href="../../../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Scythe 0.1 documentation" href="../../../index.html" />
    <link rel="up" title="Module code" href="../../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../../index.html">Scythe 0.1 documentation</a> &raquo;</li>
          <li><a href="../../index.html" accesskey="U">Module code</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <h1>Source code for scythe2.helpers.ensembl_ortho_mysql</h1><div class="highlight"><pre>
<span class="kn">import</span> <span class="nn">mysql.connector</span>

<div class="viewcode-block" id="q"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.q">[docs]</a><span class="k">def</span> <span class="nf">q</span><span class="p">(</span><span class="n">s</span><span class="p">):</span>
    <span class="k">return</span> <span class="p">(</span><span class="s">&#39;&quot;&#39;</span><span class="o">+</span><span class="n">s</span><span class="o">+</span><span class="s">&#39;&quot;&#39;</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="getGenome_Db_Ids"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.getGenome_Db_Ids">[docs]</a><span class="k">def</span> <span class="nf">getGenome_Db_Ids</span><span class="p">(</span><span class="n">specnames</span><span class="p">,</span> <span class="n">release</span><span class="p">):</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;use ensembl_compara_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">release</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;;&quot;</span> 
    <span class="n">cnx</span> <span class="o">=</span> <span class="n">mysql</span><span class="o">.</span><span class="n">connector</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="s">&#39;anonymous&#39;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&#39;ensembldb.ensembl.org&#39;</span> <span class="p">)</span>
    <span class="n">curA</span> <span class="o">=</span> <span class="n">cnx</span><span class="o">.</span><span class="n">cursor</span><span class="p">(</span><span class="n">buffered</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">specnames</span> <span class="o">=</span> <span class="p">[</span><span class="n">q</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">specnames</span><span class="p">]</span>
    <span class="n">res</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">specnames</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;SELECT genome_db_id FROM genome_db WHERE name=&#39;</span><span class="o">+</span><span class="n">s</span><span class="o">+</span><span class="s">&quot;;&quot;</span>
        <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
        <span class="n">res</span><span class="p">[</span><span class="n">s</span><span class="p">]</span> <span class="o">=</span> <span class="n">curA</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span><span class="p">(</span><span class="n">res</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="getSpecies_Set_Ids"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.getSpecies_Set_Ids">[docs]</a><span class="k">def</span> <span class="nf">getSpecies_Set_Ids</span><span class="p">(</span><span class="n">genomedb_ids</span><span class="p">,</span> <span class="n">release</span><span class="p">):</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;use ensembl_compara_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">release</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;;&quot;</span> 
    <span class="n">cnx</span> <span class="o">=</span> <span class="n">mysql</span><span class="o">.</span><span class="n">connector</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="s">&#39;anonymous&#39;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&#39;ensembldb.ensembl.org&#39;</span> <span class="p">)</span>
    <span class="n">curA</span> <span class="o">=</span> <span class="n">cnx</span><span class="o">.</span><span class="n">cursor</span><span class="p">(</span><span class="n">buffered</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">genomedb_ids</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">genomedb_ids</span><span class="p">]</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s">&quot;genome_db_id  IN (&quot;</span><span class="o">+</span> <span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">genomedb_ids</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;) &quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&quot;SELECT species_set_id,genome_db_id, count(species_set_id )&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot;from species_set  WHERE (&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="n">s</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot;) group by species_set_id having count(species_set_id) &gt; 1  ;&quot;</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">curA</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="n">res</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">res</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="getMethodLinkSpecies_Set_Ids"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.getMethodLinkSpecies_Set_Ids">[docs]</a><span class="k">def</span> <span class="nf">getMethodLinkSpecies_Set_Ids</span><span class="p">(</span><span class="n">speciesSet_ids</span><span class="p">,</span> <span class="n">release</span><span class="p">):</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;use ensembl_compara_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">release</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;;&quot;</span> 
    <span class="n">cnx</span> <span class="o">=</span> <span class="n">mysql</span><span class="o">.</span><span class="n">connector</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="s">&#39;anonymous&#39;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&#39;ensembldb.ensembl.org&#39;</span> <span class="p">)</span>
    <span class="n">curA</span> <span class="o">=</span> <span class="n">cnx</span><span class="o">.</span><span class="n">cursor</span><span class="p">(</span><span class="n">buffered</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">speciesSet_ids</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">speciesSet_ids</span><span class="p">]</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s">&quot;species_set_id  IN ( &quot;</span><span class="o">+</span> <span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">speciesSet_ids</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;) &quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&quot;SELECT method_link_species_set_id, species_set_id&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot; FROM method_link_species_set  WHERE &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="n">s</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot; ;&quot;</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">curA</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">res</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">]</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">res</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="getHomologyId"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.getHomologyId">[docs]</a><span class="k">def</span> <span class="nf">getHomologyId</span><span class="p">(</span><span class="n">method_link_species_set_ids</span><span class="p">,</span> <span class="n">release</span><span class="o">=</span><span class="mi">71</span><span class="p">):</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;use ensembl_compara_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">release</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;;&quot;</span> 
    <span class="n">cnx</span> <span class="o">=</span> <span class="n">mysql</span><span class="o">.</span><span class="n">connector</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="s">&#39;anonymous&#39;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&#39;ensembldb.ensembl.org&#39;</span> <span class="p">)</span>
    <span class="n">curA</span> <span class="o">=</span> <span class="n">cnx</span><span class="o">.</span><span class="n">cursor</span><span class="p">(</span><span class="n">buffered</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">method_link_species_set_ids</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">method_link_species_set_ids</span><span class="p">]</span>
    <span class="c">#s = &quot;method_link_species_set_id =&quot;+&quot; OR method_link_species_set_id=&quot;.join(method_link_species_set_ids)</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s">&quot;method_link_species_set_id IN (&quot;</span><span class="o">+</span><span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span> <span class="n">method_link_species_set_ids</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;) &quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&quot;SELECT homology_id, method_link_species_set_id&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot; FROM homology WHERE (&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="n">s</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot;);&quot;</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">curA</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">res</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="getHomologyMemberId"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.getHomologyMemberId">[docs]</a><span class="k">def</span> <span class="nf">getHomologyMemberId</span><span class="p">(</span><span class="n">homology_ids</span><span class="p">,</span> <span class="n">release</span><span class="p">):</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;use ensembl_compara_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">release</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;;&quot;</span> 
    <span class="n">cnx</span> <span class="o">=</span> <span class="n">mysql</span><span class="o">.</span><span class="n">connector</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="s">&#39;anonymous&#39;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&#39;ensembldb.ensembl.org&#39;</span> <span class="p">)</span>
    <span class="n">curA</span> <span class="o">=</span> <span class="n">cnx</span><span class="o">.</span><span class="n">cursor</span><span class="p">(</span><span class="n">buffered</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">homology_ids</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">homology_ids</span><span class="p">]</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s">&quot;homology_id =&quot;</span><span class="o">+</span><span class="s">&quot; OR homology_id=&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">homology_ids</span><span class="p">)</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&quot;SELECT homology_member.member_id, member.member_id,member.stable_id, homology_id &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot; FROM homology_member LEFT JOIN  member ON &quot;</span>
    <span class="n">cmd</span><span class="o">+=</span> <span class="s">&quot;(homology_member.member_id = member.member_id) WHERE (&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="n">s</span>
    <span class="n">cmd</span> <span class="o">+=</span> <span class="s">&quot;);&quot;</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">curA</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">res</span><span class="p">)</span>
</div>
<div class="viewcode-block" id="fetch1to1orthologs"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.fetch1to1orthologs">[docs]</a><span class="k">def</span> <span class="nf">fetch1to1orthologs</span><span class="p">(</span><span class="n">method_link_species_set_ids</span><span class="p">,</span><span class="n">release</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return (homology_id, genome_db_id, stable_id)&quot;&quot;&quot;</span>
    <span class="n">method_link_species_set_ids</span><span class="o">=</span><span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">s</span><span class="p">)</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">method_link_species_set_ids</span><span class="p">]</span>

    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&#39;use ensembl_compara_&#39;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">release</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;;&quot;</span> 
    <span class="n">cnx</span> <span class="o">=</span> <span class="n">mysql</span><span class="o">.</span><span class="n">connector</span><span class="o">.</span><span class="n">connect</span><span class="p">(</span><span class="n">user</span><span class="o">=</span><span class="s">&#39;anonymous&#39;</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s">&#39;ensembldb.ensembl.org&#39;</span> <span class="p">)</span>
    <span class="n">curA</span> <span class="o">=</span> <span class="n">cnx</span><span class="o">.</span><span class="n">cursor</span><span class="p">(</span><span class="n">buffered</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">s</span> <span class="o">=</span> <span class="s">&quot;method_link_species_set.method_link_species_set_id IN (&quot;</span><span class="o">+</span> <span class="s">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">method_link_species_set_ids</span><span class="p">)</span><span class="o">+</span><span class="s">&quot;) &quot;</span>
    <span class="n">cmd</span> <span class="o">=</span> <span class="s">&quot; SELECT DISTINCT homology.homology_id,  member.genome_db_id,  member.stable_id from&quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot; homology, homology_member, member, species_set, method_link_species_set &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot; WHERE ( &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="n">s</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&#39;AND homology.description=&quot;ortholog_one2one&quot;&#39;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot;AND &quot;</span><span class="o">+</span><span class="n">s</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot;AND species_set.species_set_id = method_link_species_set.species_set_id &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot;AND homology.method_link_species_set_id = method_link_species_set.method_link_species_set_id &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot;AND homology_member.member_id=member.member_id &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot;AND homology_member.homology_id=homology.homology_id &quot;</span>
    <span class="n">cmd</span> <span class="o">+=</span><span class="s">&quot;)&quot;</span>

    <span class="n">curA</span><span class="o">.</span><span class="n">execute</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">curA</span><span class="o">.</span><span class="n">fetchall</span><span class="p">()</span>
    <span class="n">curA</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">res</span><span class="p">)</span>


</div>
<div class="viewcode-block" id="makeTable"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.makeTable">[docs]</a><span class="k">def</span> <span class="nf">makeTable</span><span class="p">(</span><span class="n">genomeDB_ids</span><span class="p">,</span> <span class="n">orthoSpec2stableIds</span><span class="p">,</span> <span class="n">orthoIds</span><span class="p">,</span> <span class="n">outfile</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">SPECID</span>
    <span class="k">global</span> <span class="n">OUTFILES</span>
    <span class="n">res</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">specset</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">orthoIds</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">genomeDB_ids</span><span class="p">:</span>
            <span class="k">if</span> <span class="p">(</span><span class="n">o</span><span class="p">,</span><span class="n">i</span><span class="p">)</span> <span class="ow">in</span> <span class="n">orthoSpec2stableIds</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
                    <span class="n">res</span><span class="p">[</span><span class="n">o</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">res</span><span class="p">[</span><span class="n">o</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">((</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">1</span><span class="p">]))</span> <span class="ow">in</span> <span class="n">specset</span><span class="p">:</span>
                 <span class="n">specset</span><span class="p">[(</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">1</span><span class="p">])]</span><span class="o">=</span><span class="p">[(</span><span class="n">orthoSpec2stableIds</span><span class="p">[(</span><span class="n">r</span><span class="p">,</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">0</span><span class="p">])],</span><span class="n">orthoSpec2stableIds</span><span class="p">[(</span><span class="n">r</span><span class="p">,</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="p">)])</span> <span class="p">]</span>
    
        <span class="k">else</span><span class="p">:</span>
             <span class="n">specset</span><span class="p">[(</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">1</span><span class="p">])]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">orthoSpec2stableIds</span><span class="p">[(</span><span class="n">r</span><span class="p">,</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">0</span><span class="p">])],</span><span class="n">orthoSpec2stableIds</span><span class="p">[(</span><span class="n">r</span><span class="p">,</span><span class="n">res</span><span class="p">[</span><span class="n">r</span><span class="p">][</span><span class="mi">1</span><span class="p">])]))</span>
    
    <span class="k">print</span><span class="p">(</span><span class="n">specset</span><span class="p">)</span> 
    <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">specset</span><span class="p">:</span>
        <span class="n">st</span> <span class="o">=</span> <span class="n">SPECID</span><span class="p">[</span><span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">+</span><span class="s">&quot;__&quot;</span><span class="o">+</span><span class="n">SPECID</span><span class="p">[</span><span class="n">s</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
        <span class="k">print</span><span class="p">(</span><span class="n">st</span><span class="p">)</span>
        <span class="c">#try:</span>
        <span class="c">#    if not outfile:</span>
        <span class="n">outfile</span> <span class="o">=</span> <span class="n">st</span>
        <span class="n">out</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span><span class="s">&quot;w&quot;</span><span class="p">)</span>
        <span class="c">#    else:</span>
        <span class="c">#        out = open(outfile,&quot;w&quot;)</span>
        <span class="c">#except IOError as e:</span>
        <span class="c">#    print(e)</span>
        <span class="c">#    exit(1)</span>
        <span class="k">print</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span><span class="s">&quot;OUTFILE&quot;</span><span class="p">)</span>
        <span class="n">OUTFILES</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">outfile</span><span class="p">)</span>
        <span class="n">tuplist</span> <span class="o">=</span> <span class="n">specset</span><span class="p">[</span><span class="n">s</span><span class="p">]</span>
        <span class="k">for</span> <span class="n">tup</span> <span class="ow">in</span> <span class="n">tuplist</span><span class="p">:</span>
            <span class="n">tup</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">tup</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">+</span><span class="s">&quot;</span><span class="se">\t</span><span class="s">&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">tup</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
            <span class="k">print</span><span class="p">(</span><span class="n">tup</span><span class="p">)</span>
            <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">tup</span><span class="p">)</span>
            <span class="n">out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&quot;</span><span class="se">\n</span><span class="s">&quot;</span><span class="p">)</span>
        <span class="n">out</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="c">### glob var ###</span></div>
<span class="n">SPECID</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">OUTFILES</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c">################</span>
<div class="viewcode-block" id="fetchOrthoFromMySQL"><a class="viewcode-back" href="../../../Modules.html#scythe2.helpers.ensembl_ortho_mysql.fetchOrthoFromMySQL">[docs]</a><span class="k">def</span> <span class="nf">fetchOrthoFromMySQL</span><span class="p">(</span><span class="n">specieslist</span> <span class="o">=</span> <span class="p">[</span><span class="s">&quot;homo_sapiens&quot;</span><span class="p">,</span><span class="s">&quot;pan_troglodytes&quot;</span><span class="p">,</span><span class="s">&quot;mus_musculus&quot;</span><span class="p">],</span> <span class="n">release</span><span class="o">=</span><span class="mi">74</span><span class="p">,</span> <span class="n">outfile</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="n">specieslist</span><span class="p">)</span>
    <span class="n">name2genomeDB_ids</span> <span class="o">=</span> <span class="n">getGenome_Db_Ids</span><span class="p">(</span><span class="n">specieslist</span><span class="p">,</span> <span class="n">release</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">name2genomeDB_ids</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">name2genomeDB_ids</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">SPECID</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">v</span>
        <span class="n">SPECID</span><span class="p">[</span><span class="n">v</span><span class="p">]</span> <span class="o">=</span> <span class="n">k</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">orthoIds</span> <span class="o">=</span> <span class="p">[]</span> 
    <span class="n">orthoSpec2stableIds</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    <span class="n">members</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
    
    <span class="n">genomeDB_ids</span> <span class="o">=</span> <span class="n">name2genomeDB_ids</span><span class="o">.</span><span class="n">values</span><span class="p">()</span>
    <span class="k">print</span><span class="p">(</span><span class="n">genomeDB_ids</span><span class="p">)</span>
    <span class="n">speciesSet_ids</span> <span class="o">=</span> <span class="n">getSpecies_Set_Ids</span><span class="p">(</span><span class="n">genomeDB_ids</span><span class="p">,</span><span class="n">release</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">speciesSet_ids</span><span class="p">,</span><span class="s">&quot;specsetids&quot;</span><span class="p">)</span>
    <span class="n">methodLinkSpeciesSet_ids</span> <span class="o">=</span> <span class="n">getMethodLinkSpecies_Set_Ids</span><span class="p">(</span><span class="n">speciesSet_ids</span><span class="p">,</span> <span class="n">release</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">methodLinkSpeciesSet_ids</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">fetch1to1orthologs</span><span class="p">(</span><span class="n">methodLinkSpeciesSet_ids</span><span class="p">,</span> <span class="n">release</span><span class="o">=</span><span class="n">release</span><span class="p">)</span>
    
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
        <span class="k">print</span><span class="p">(</span><span class="s">&quot;res&quot;</span><span class="p">,</span><span class="n">r</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="ow">in</span> <span class="n">orthoSpec2stableIds</span><span class="p">:</span>
            <span class="c">#r[0] is shared between orthologous genes from different species</span>
            <span class="c">#r[1] is the genome_id </span>
            <span class="c">#r[2] is the stable_id of the gene</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">orthoIds</span><span class="p">:</span>
                <span class="n">orthoIds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">orthoSpec2stableIds</span><span class="p">[(</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span><span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">])]</span><span class="o">=</span><span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="c">#[0]</span>
    
    
    <span class="n">makeTable</span><span class="p">(</span><span class="n">genomeDB_ids</span><span class="p">,</span> <span class="n">orthoSpec2stableIds</span><span class="p">,</span> <span class="n">orthoIds</span><span class="p">,</span> <span class="n">outfile</span><span class="p">)</span>
    <span class="k">return</span><span class="p">(</span><span class="n">OUTFILES</span><span class="p">)</span>
<span class="c">#fetchOrthoFromMySQL()</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li><a href="../../../index.html">Scythe 0.1 documentation</a> &raquo;</li>
          <li><a href="../../index.html" >Module code</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Janina Mass.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>