% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{Scythe Documentation}
\date{May 22, 2014}
\release{0.1}
\author{Janina Mass}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


Scythe -- Selection of Conserved Transcripts by Homology Evaluation

This documentation is under construction!

Contents:


\chapter{Introduction}
\label{Introduction:introduction}\label{Introduction::doc}\label{Introduction:welcome-to-scythe-s-documentation}
Scythe picks best matching transcripts for one-to-one orthologous genes from two or more species.
The goal is to provide the best, i.e. most homologous set of sequences, for a subsequent multiple sequence alignment in order
to minimize sources for misalignment in an automated fashion.
Scythe will perform pairwise global alignments using the Needleman-Wunsch algorithm {[}NE1970{]} as implemented in needleall as part of the EMBOSS package {[}EMB2000{]}.

Please see the {\hyperref[Tutorial:tutorial]{\emph{Tutorial}}} on how to use Scythe.
Important! needleall seems to be broken in EMBOSS 6.6.0.0; this is working with needleall from EMBOSS 6.4.0.0.


\section{Algorithms and Application}
\label{Introduction:algorithms-and-application}\label{Introduction:algo}
Scythe performs pairwise global alignments {[}NEED1970{]} to measure the similarity between transcripts.
Transcripts are only compared between species.

There are three main strategies implemented in Scythe to deal with the sequences after scoring:
A single-linkage approach either starting with reference transcripts as seed ({\hyperref[Introduction:sl-ref]{sl\_ref}}) and adding best matching sequences from non-reference species
or starting out with the best matching transcript pair for a gene ({\hyperref[Introduction:sl-glob]{sl\_glob}}).
A \emph{reference species} is defined as a species that has only one transcript (\emph{reference transcript}) for a gene.
Every gene is processed individually, a \emph{reference species} is only local to a gene and is automatically derived from the input data.
Alternatively, the maximum-sum ({\hyperref[Introduction:mx-sum]{mx\_sum}}) approach calculates the score for all transcript pairings between the species and return a maximum-scoring set.
Please note that this approach might not be feasable for large data sets.


\subsection{mx\_sum}
\label{Introduction:id1}\label{Introduction:mx-sum}
mx\_sum (maximum-sum)
max sum algorithm returns an optimal solution for the problem of score maximization.
There are, however, scenarios where this would not represent the desired result: In cases of single-transcript outliers,
their pairwise score to all other sequences is always taken into account and may favor sequences that are less dissimilar to the outlier.


\subsection{sl\_ref}
\label{Introduction:sl-ref}\label{Introduction:id2}
sl\_ref(single-linkage-reference)...
It works well in cases where the user wants to
focus on finding similar gene models given one reference species, e. g. when the user
has already decided on a splice variant in one species.


\subsection{sl\_glob}
\label{Introduction:sl-glob}
sl\_glob (single-linkage-global) due to the single linkage selection approach, the algorithm starting with the best match-
ing pair should provide a good starting point in absence of a reference gene model and
may be able circumvent outliers.


\subsection{bf\_msa}
\label{Introduction:bf-msa}
!todo
brute force all vs all msa comparison using mafft


\chapter{Tutorial}
\label{Tutorial::doc}\label{Tutorial:tutorial}\label{Tutorial:id1}

\section{Installation}
\label{Tutorial:installation}

\subsection{Requirements}
\label{Tutorial:requirements}\begin{itemize}
\item {} 
Python 3

\end{itemize}


\subsubsection{python modules}
\label{Tutorial:python-modules}\begin{itemize}
\item {} 
configparser

\item {} 
mysql-connector-python

\item {} 
httplib2

\end{itemize}

\begin{Verbatim}[commandchars=\\\{\}]
pip install \PYGZhy{}\PYGZhy{}allow\PYGZhy{}external mysql\PYGZhy{}connector\PYGZhy{}python mysql\PYGZhy{}connector\PYGZhy{}python
pip install httplib2
pip install configparser
\end{Verbatim}

In case you want to work using a virtual environment switch to that environment first (see {\hyperref[Tutorial:install-virt]{\emph{Install Scythe in a virtual environment}}}).


\subsubsection{external software}
\label{Tutorial:external-software}\begin{itemize}
\item {} 
needleall \href{http://emboss.sourceforge.net/}{http://emboss.sourceforge.net/}

\end{itemize}

!Important: needleall from (the ubuntu package for) EMBOSS 6.6.0.0 causes problems. This program was tested and works with needleall from EMBOSS 6.4.0.0.

Install emboss on debian-based systems:

\begin{Verbatim}[commandchars=\\\{\}]
sudo apt\PYGZhy{}get install emboss
\end{Verbatim}


\subsection{Install Scythe in a virtual environment}
\label{Tutorial:install-scythe-in-a-virtual-environment}\label{Tutorial:install-virt}
Install virtualenv via pip:

\begin{Verbatim}[commandchars=\\\{\}]
pip install virtualenv
\end{Verbatim}

Create a virtual environment that uses Python 3 and activate:

\begin{Verbatim}[commandchars=\\\{\}]
virtualenv \PYGZhy{}p /usr/bin/python3 python3env
source python3env/bin/activate
\end{Verbatim}

Finally, install the scythe package:

\begin{Verbatim}[commandchars=\\\{\}]
(python3env)you@host:\PYGZti{}\PYGZdl{} pip install scythe
\end{Verbatim}


\subsection{Without virtual environment}
\label{Tutorial:without-virtual-environment}
Via pip:

\begin{Verbatim}[commandchars=\\\{\}]
pip install scythe
\end{Verbatim}

From source:

\begin{Verbatim}[commandchars=\\\{\}]
wget TODO
tar xvf TODO
cd TODO/DIR
python setup.py install
\end{Verbatim}


\section{Using Scythe}
\label{Tutorial:usingscythe}\label{Tutorial:using-scythe}
There are three ways to use Scythe: GUI, command line parameters, or configuration file.

The GUI allows you to directy obtain data from ENSEMBL, have it converted and processed by Scythe.

If you are using local files, make sure you have directories prepared that contain the \emph{loc} and fasta files, and that you have a \emph{grp} file ready.
(See {\hyperref[Converters:converters]{\emph{Converters}}} on how to convert other formats to \emph{loc} or \emph{grp}.)

Read more about configuration files under {\hyperref[ConfigurationFiles:configurationfiles]{\emph{Configuration Files}}}.


\subsection{Graphical user interface}
\label{Tutorial:gui}\label{Tutorial:graphical-user-interface}
Call the GUI via

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{scythe}\PYG{o}{\PYGZhy{}}\PYG{n}{gui}\PYG{o}{.}\PYG{n}{py}
\end{Verbatim}

or with configuration file \emph{conf.py}

\begin{Verbatim}[commandchars=\\\{\}]
scythe\PYGZhy{}gui.py conf.scy
\end{Verbatim}

Please note that Scythe will write intermediate files to the current working directory. You might want to create a new directory for each run beforehand.
Data that was already downloaded from ENSEMBL will not be downloaded again. If you want to reuse this data but try different parameters, just change the output directory (via the GUI) and start Scythe from the same working directory.


\subsection{Command line interface}
\label{Tutorial:command-line-interface}\label{Tutorial:cli}
\begin{Verbatim}[commandchars=\\\{\}]
    \PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}
    \PYGZsh{} scythe.py v0.1                     \PYGZsh{}
    \PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}
  usage:
     scythe.py \PYGZhy{}i DIR \PYGZhy{}g .grpFILE \PYGZhy{}\PYGZhy{}cleanup

  usage with configuration file:
     scythe.py \PYGZhy{}\PYGZhy{}config configuration.scy

  general options:
    \PYGZhy{}C, \PYGZhy{}\PYGZhy{}config                     use configuration file instead of
                                     command line parameters
    \PYGZhy{}c, \PYGZhy{}\PYGZhy{}cleanup                    remove temporary files when done
    \PYGZhy{}h, \PYGZhy{}\PYGZhy{}help                       prints this
    \PYGZhy{}i, \PYGZhy{}\PYGZhy{}in\PYGZus{}dir=DIR                 folder w/ subfolders \PYGZdq{}fa\PYGZdq{} and \PYGZdq{}loc\PYGZdq{}
    \PYGZhy{}r, \PYGZhy{}\PYGZhy{}sl\PYGZus{}ref                     find best matches to reference
    \PYGZhy{}g, \PYGZhy{}\PYGZhy{}sl\PYGZus{}glob                    best scoring pair as seed
    \PYGZhy{}m, \PYGZhy{}\PYGZhy{}mx\PYGZus{}sum                     optimize sum of pairwise scores
    \PYGZhy{}o, \PYGZhy{}\PYGZhy{}out\PYGZus{}dir=DIR                output directory [default:./]
    \PYGZhy{}v, \PYGZhy{}\PYGZhy{}verbose                    be wordy

  alignment options:
     \PYGZhy{}O, \PYGZhy{}\PYGZhy{}gap\PYGZus{}open=FLOAT           needleall gap opening cost [default 10]
     \PYGZhy{}E, \PYGZhy{}\PYGZhy{}gap\PYGZus{}extend=FLOAT         needleall gap extension cost

  fasta options:
    \PYGZhy{}d, \PYGZhy{}\PYGZhy{}delim=STRING               split fasta headers at STRING
    \PYGZhy{}a, \PYGZhy{}\PYGZhy{}asID=INT                   use INTth part of fasta header as transcript\PYGZhy{}ID
                                     (default:0)
  debug options:
    \PYGZhy{}s, \PYGZhy{}\PYGZhy{}stop\PYGZus{}after=NUM             stop after NUM groups

  further help:
    Please see documentation in the \PYGZsq{}docs\PYGZsq{}\PYGZhy{}directory.
\end{Verbatim}

Please note that you cannot automatically download sequences from ENSEMBL via the command line interface.


\chapter{Format}
\label{Format:id1}\label{Format::doc}\label{Format:format}
The formats have a very simple, human readable structure allowing users to also manually modify or create them.
For locus-gene model identifier relations: {\hyperref[Format:loc-format]{\emph{loc format}}}
For grouping of orthologous genes {\hyperref[Format:grp-format]{\emph{grp format}}}


\section{loc format}
\label{Format:loc-format}\label{Format:id2}
Each row of a \emph{loc} file represents a set of gene models (GM) for a gene locus (L).
Rows start with a the gene locus ID followed by its gene model identifiers.
The GM in the second column in the file is treated as the reference form for its gene locus,
i.e. in a case of a tie, this gene model is preferred.
The columns are tab-separated.

File \emph{Spec0.loc}:

\begin{Verbatim}[commandchars=\\\{\}]
Sp0L0       Sp0L0GM0        Sp0L0GM1        . . . Sp0L0GMjLF
Sp0L1       Sp1L0GM0        Sp0L1GM1        . . . Sp0L1GMkLF
.
.
.
Sp0Lm       Sp1LmGM0        Sp0LmGM1        . . . Sp0LmGMlLF
\end{Verbatim}


\section{grp format}
\label{Format:grp-format}\label{Format:id3}
Each row of a \emph{grp} file represents an orthologous group.
Rows start with a unique group ID followed by orthologous gene identifiers for the respective species.
Rows don't need to have the same number of columns. The gene loci order within a row is irrelevant.
The programm will keep the identifiers for its output.
The columns are tab-separated.

File \emph{orthogroups.grp} (Sp: species, L: orthologous gene loci):

\begin{Verbatim}[commandchars=\\\{\}]
0   SpaLw   SpbLw   SpcLw   . . . SpzLw
1   SpaLx   SpbLx   SpcLx   . . .
.
.
.
N   SpaLz   SpbLz   SpcLz   . . .
\end{Verbatim}


\chapter{Converters}
\label{Converters::doc}\label{Converters:converters}\label{Converters:id1}
Scythe uses simple, human readable  formats to store gene-transcript and ortholog information.
See also {\hyperref[Format:format]{\emph{Format}}}.


\section{loc}
\label{Converters:loc}
Scripts to convert the following formats to \emph{loc} format are included:
\begin{itemize}
\item {} 
gff3

\item {} 
tab-separated (eg ENSEMBL BioMart)

\end{itemize}

run

\begin{Verbatim}[commandchars=\\\{\}]
scythe\PYGZus{}loc\PYGZus{}gff.py \PYGZhy{}f GFF
\end{Verbatim}

or

\begin{Verbatim}[commandchars=\\\{\}]
scythe\PYGZus{}loc\PYGZus{}tsv.py \PYGZhy{}f FILE.tsv
\end{Verbatim}

See also: {\hyperref[Format:loc-format]{\emph{loc format}}}.


\section{grp}
\label{Converters:grp}
Please note that the \emph{grp} converters  need a concatenated \emph{loc} file in addition to the orthology information.
Scripts to convert the following formats to \emph{grp} are included:
\begin{itemize}
\item {} 
orthomcl

\item {} 
proteinortho

\item {} 
tab-separated (eg ENSEMBL BioMart)

\end{itemize}

run

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{scythe\PYGZus{}grp\PYGZus{}orthomcl}\PYG{o}{.}\PYG{n}{py}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{scythe\PYGZus{}grp\PYGZus{}proteinortho}\PYG{o}{.}\PYG{n}{py}
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{scythe\PYGZus{}grp\PYGZus{}tsv}\PYG{o}{.}\PYG{n}{py}
\end{Verbatim}

See also {\hyperref[Format:grp-format]{\emph{grp format}}}.
\phantomsection\label{Converters:module-scythe2.convert}\index{scythe2.convert (module)}\begin{quote}\begin{description}
\item[{undoc-members}] \leavevmode
\end{description}\end{quote}

\begin{Verbatim}[commandchars=\\\{\}]
    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{} loc output format \PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}

    LOCUS0	TRANSCRIPT0\PYGZus{}0	TRANSCRIPT0\PYGZus{}1	...	TRANSCRIPT0\PYGZus{}n
    LOCUS1	TRANSCRIPT1\PYGZus{}0	...	TRANSCRIPT1\PYGZus{}m
    .
    .
    .
    LOCUSk	TRANSCRIPTk\PYGZus{}0	...TRANSCRIPTk\PYGZus{}l

    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}


    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{} gff version 3 input format \PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}
    example (tab is shown as \PYGZbs{}t):

    \PYGZsh{}\PYGZsh{}gff\PYGZhy{}version 3
    L1i\PYGZbs{}texample\PYGZbs{}tgene\PYGZbs{}t1000\PYGZbs{}t9000\PYGZbs{}t.\PYGZbs{}t+\PYGZbs{}t.\PYGZbs{}tID=L1;Name=L1;Note=example
    L1.a\PYGZbs{}texample\PYGZbs{}tmRNA\PYGZbs{}t1000\PYGZbs{}t9000\PYGZbs{}t.\PYGZbs{}t+\PYGZbs{}t.\PYGZbs{}tID=L1.a;Parent=L1;Name=L1.a


    Note that this script only relies on the \PYGZdq{}ID\PYGZdq{} and \PYGZdq{}Parent\PYGZdq{} tags,
    \PYGZdq{}Name\PYGZdq{} will be ignored. If \PYGZdq{}longest\PYGZdq{} is specified (eg phytozome)
    and ==\PYGZdq{}1\PYGZdq{}, the transcript will be placed on the first position
    for its gene.
    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}

    

    \PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}
    \PYGZsh{}  scythe\PYGZus{}loc\PYGZus{}gff.py  \PYGZhy{}f FILE.gff3  \PYGZsh{}
    \PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}

    \PYGZhy{}f, \PYGZhy{}\PYGZhy{}file=gff3\PYGZus{}FILE
    \PYGZhy{}o, \PYGZhy{}\PYGZhy{}output=FILE        output file [default: gff3\PYGZus{}FILE.loc]
    \PYGZhy{}h, \PYGZhy{}\PYGZhy{}help               prints this
    \PYGZhy{}H, \PYGZhy{}\PYGZhy{}HELP               show help on format
\end{Verbatim}

\begin{Verbatim}[commandchars=\\\{\}]
    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{} loc output format \PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}

    LOCUS0	TRANSCRIPT0\PYGZus{}0	TRANSCRIPT0\PYGZus{}1	...	TRANSCRIPT0\PYGZus{}n
    LOCUS1	TRANSCRIPT1\PYGZus{}0	...	TRANSCRIPT1\PYGZus{}m
    .
    .
    .
    LOCUSk	TRANSCRIPTk\PYGZus{}0	...TRANSCRIPTk\PYGZus{}l

    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}

    

    example ensembl query:

    http://www.ensembl.org/biomart/
    \PYGZlt{}?xml version=\PYGZdq{}1.0\PYGZdq{} encoding=\PYGZdq{}UTF\PYGZhy{}8\PYGZdq{}?\PYGZgt{}
    \PYGZlt{}!DOCTYPE Query\PYGZgt{}
    \PYGZlt{}Query  virtualSchemaName = \PYGZdq{}default\PYGZdq{} formatter = \PYGZdq{}TSV\PYGZdq{}
    header = \PYGZdq{}0\PYGZdq{} uniqueRows = \PYGZdq{}0\PYGZdq{} count = \PYGZdq{}\PYGZdq{} datasetConfigVersion = \PYGZdq{}0.6\PYGZdq{} \PYGZgt{}
    \PYGZlt{}Dataset name = \PYGZdq{}hsapiens\PYGZus{}gene\PYGZus{}ensembl\PYGZdq{} interface = \PYGZdq{}default\PYGZdq{} \PYGZgt{}
    \PYGZlt{}Filter name = \PYGZdq{}biotype\PYGZdq{} value = \PYGZdq{}protein\PYGZus{}coding\PYGZdq{}/\PYGZgt{}
    \PYGZlt{}Attribute name = \PYGZdq{}ensembl\PYGZus{}gene\PYGZus{}id\PYGZdq{} /\PYGZgt{}
    \PYGZlt{}Attribute name = \PYGZdq{}ensembl\PYGZus{}transcript\PYGZus{}id\PYGZdq{} /\PYGZgt{}
    \PYGZlt{}Attribute name = \PYGZdq{}ensembl\PYGZus{}peptide\PYGZus{}id\PYGZdq{} /\PYGZgt{}
    \PYGZlt{}Attribute name = \PYGZdq{}cds\PYGZus{}length\PYGZdq{} /\PYGZgt{}
    \PYGZlt{}/Dataset\PYGZgt{}
    \PYGZlt{}/Query\PYGZgt{}
    

    \PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}
    \PYGZsh{}  scythe\PYGZus{}loc\PYGZus{}tsv.py      \PYGZsh{}
    \PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}\PYGZsh{}
    \PYGZhy{}f, \PYGZhy{}\PYGZhy{}file=ENSEMBLBioMart.tsv
                              format: 1st column: gene id, 2nd column:transcript id,
                                      3rd column: peptide id, 4th column: cds length;
                                      gene ids can occur multiple times
    \PYGZhy{}c, \PYGZhy{}\PYGZhy{}custom=COLx,COLy,COLz,...   COLi in [\PYGZdq{}gene\PYGZdq{},\PYGZdq{}transcript\PYGZdq{}, \PYGZdq{}protein\PYGZdq{}, \PYGZdq{}length\PYGZdq{}]
                                      Use this if your file is different from the described biomart output.
                                      \PYGZdq{}cds\PYGZus{}length\PYGZdq{} is optional but recommended, at least one of
                                      [\PYGZdq{}transcript\PYGZdq{}, \PYGZdq{}protein\PYGZdq{}] need to be included
    \PYGZhy{}o, \PYGZhy{}\PYGZhy{}output=FILE         output file [default: ENSEMBLEBioMart.tsv.loc]
    \PYGZhy{}h, \PYGZhy{}\PYGZhy{}help                prints this
    \PYGZhy{}H, \PYGZhy{}\PYGZhy{}HELP                show help on format
    \PYGZsh{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZhy{}\PYGZsh{}
\end{Verbatim}


\chapter{Configuration Files}
\label{ConfigurationFiles:configuration-files}\label{ConfigurationFiles::doc}\label{ConfigurationFiles:configurationfiles}
The configuration files allow you save your settings/restore your settings.
Most entries should be self-explanatory.


\section{Example file}
\label{ConfigurationFiles:example-file}
\begin{Verbatim}[commandchars=\\\{\}]
[Mode]
use\PYGZus{}ensembl=yes
use\PYGZus{}local\PYGZus{}files=no

[Paths]
fasta\PYGZus{}directory=/home/nin/py3env/fa/
loc\PYGZus{}directory=/home/nin/py3env/loc/
grp\PYGZus{}file=/home/nin/py3env/magohopopafe.shared\PYGZus{}by\PYGZus{}all.grp
output\PYGZus{}directory=/home/nin/py3env

[Cleanup]
clean\PYGZus{}up\PYGZus{}directories=yes

[Run\PYGZus{}options]
max\PYGZus{}threads=1
split\PYGZus{}input=1

[Penalties]
gap\PYGZus{}open\PYGZus{}cost=10
gap\PYGZus{}extend\PYGZus{}cost=0.5
substitution\PYGZus{}matrix=EBLOSUM62

[Algorithm]
use\PYGZus{}sl\PYGZus{}glob=unset
use\PYGZus{}sl\PYGZus{}ref=unset
use\PYGZus{}mx\PYGZus{}sum=yes

[Fasta\PYGZus{}header]
fasta\PYGZus{}header\PYGZus{}delimiter=\PYGZdq{} \PYGZdq{}
fasta\PYGZus{}header\PYGZus{}part=0
\end{Verbatim}


\section{{[}Fasta\_header{]}}
\label{ConfigurationFiles:fasta-header}
In case the ids in your \emph{loc} file are truncated or your fasta headers carry information in addition to the sequence identifier, you can define where to split the fasta header.

Example:

\begin{Verbatim}[commandchars=\\\{\}]
\PYGZgt{}other:xyz\PYGZus{}sequenceID\PYGZus{}length:123\PYGZus{}species:x
MYQVLQAYDWKYLHDNHDCNFQVGGADQLGNI...
\end{Verbatim}

and:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{p}{[}\PYG{n}{Fasta\PYGZus{}header}\PYG{p}{]}
\PYG{n}{fasta\PYGZus{}header\PYGZus{}delimiter} \PYG{o}{=} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{\PYGZus{}}\PYG{l+s}{\PYGZdq{}}
\PYG{n}{fasta\PYGZus{}header\PYGZus{}part} \PYG{o}{=} \PYG{l+m+mi}{1}
\end{Verbatim}

Would result in taking only \emph{sequenceID} into account.


\section{{[}Run\_options{]}}
\label{ConfigurationFiles:run-options}
Multi core support is not yet implemented.


\section{{[}Algorithm{]}}
\label{ConfigurationFiles:algorithm}
See {\hyperref[Introduction:algo]{\emph{Algorithms and Application}}}


\chapter{Modules}
\label{Modules:modules}\label{Modules::doc}

\section{convert}
\label{Modules:convert}\label{Modules:module-scythe2.convert}\index{scythe2.convert (module)}\phantomsection\label{Modules:module-scythe2.convert.scythe_loc_gff}\index{scythe2.convert.scythe\_loc\_gff (module)}\index{checkGff3() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.checkGff3}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{checkGff3}}{\emph{infile}}{}
\end{fulllineitems}

\index{formatHelp() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.formatHelp}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{formatHelp}}{}{}
\end{fulllineitems}

\index{main() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.main}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{main}}{}{}
\end{fulllineitems}

\index{read\_gene\_mrna() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.read_gene_mrna}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{read\_gene\_mrna}}{\emph{infile}}{}
\end{fulllineitems}

\index{read\_gff2loc() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.read_gff2loc}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{read\_gff2loc}}{\emph{infile}, \emph{outfile}}{}
\end{fulllineitems}

\index{read\_gff\_attrib() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.read_gff_attrib}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{read\_gff\_attrib}}{\emph{infile}, \emph{attrib}}{}
\end{fulllineitems}

\index{usage() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.usage}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{usage}}{}{}
Print usage.

\end{fulllineitems}

\index{writeLoc() (in module scythe2.convert.scythe\_loc\_gff)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_gff.writeLoc}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_gff.}\bfcode{writeLoc}}{\emph{genesDct}, \emph{outfile}}{}
\end{fulllineitems}

\phantomsection\label{Modules:module-scythe2.convert.scythe_loc_tsv}\index{scythe2.convert.scythe\_loc\_tsv (module)}\index{WrongNumberOfColumnsException}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.WrongNumberOfColumnsException}\pysigline{\strong{exception }\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{WrongNumberOfColumnsException}}
\end{fulllineitems}

\index{formatHelp() (in module scythe2.convert.scythe\_loc\_tsv)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.formatHelp}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{formatHelp}}{}{}
\end{fulllineitems}

\index{howTo() (in module scythe2.convert.scythe\_loc\_tsv)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.howTo}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{howTo}}{}{}
Print an ENSEMBLE example query

\end{fulllineitems}

\index{main() (in module scythe2.convert.scythe\_loc\_tsv)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.main}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{main}}{}{}
\end{fulllineitems}

\index{readTsv() (in module scythe2.convert.scythe\_loc\_tsv)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.readTsv}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{readTsv}}{\emph{infile=None}}{}
\end{fulllineitems}

\index{usage() (in module scythe2.convert.scythe\_loc\_tsv)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.usage}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{usage}}{}{}
\end{fulllineitems}

\index{writeLoc() (in module scythe2.convert.scythe\_loc\_tsv)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.convert.scythe_loc_tsv.writeLoc}\pysiglinewithargsret{\code{scythe2.convert.scythe\_loc\_tsv.}\bfcode{writeLoc}}{\emph{genesDct}, \emph{outfile}}{}
\end{fulllineitems}



\section{gui}
\label{Modules:module-scythe2.gui}\label{Modules:gui}\index{scythe2.gui (module)}

\section{helpers}
\label{Modules:module-scythe2.helpers}\label{Modules:helpers}\index{scythe2.helpers (module)}\phantomsection\label{Modules:module-scythe2.helpers.ensembl2grp}\index{scythe2.helpers.ensembl2grp (module)}\index{main() (in module scythe2.helpers.ensembl2grp)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl2grp.main}\pysiglinewithargsret{\code{scythe2.helpers.ensembl2grp.}\bfcode{main}}{}{}
\end{fulllineitems}

\index{readTsvFiles() (in module scythe2.helpers.ensembl2grp)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl2grp.readTsvFiles}\pysiglinewithargsret{\code{scythe2.helpers.ensembl2grp.}\bfcode{readTsvFiles}}{\emph{listoftsv}, \emph{outfile}}{}
\end{fulllineitems}

\index{usage() (in module scythe2.helpers.ensembl2grp)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl2grp.usage}\pysiglinewithargsret{\code{scythe2.helpers.ensembl2grp.}\bfcode{usage}}{}{}
\end{fulllineitems}

\phantomsection\label{Modules:module-scythe2.helpers.ensembl_ortho_mysql}\index{scythe2.helpers.ensembl\_ortho\_mysql (module)}\index{fetch1to1orthologs() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.fetch1to1orthologs}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{fetch1to1orthologs}}{\emph{method\_link\_species\_set\_ids}, \emph{release}}{}
Return (homology\_id, genome\_db\_id, stable\_id)

\end{fulllineitems}

\index{fetchOrthoFromMySQL() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.fetchOrthoFromMySQL}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{fetchOrthoFromMySQL}}{\emph{specieslist={[}'homo\_sapiens', `pan\_troglodytes', `mus\_musculus'{]}, release=74, outfile=None}}{}
\end{fulllineitems}

\index{getGenome\_Db\_Ids() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.getGenome_Db_Ids}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{getGenome\_Db\_Ids}}{\emph{specnames}, \emph{release}}{}
\end{fulllineitems}

\index{getHomologyId() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.getHomologyId}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{getHomologyId}}{\emph{method\_link\_species\_set\_ids}, \emph{release=71}}{}
\end{fulllineitems}

\index{getHomologyMemberId() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.getHomologyMemberId}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{getHomologyMemberId}}{\emph{homology\_ids}, \emph{release}}{}
\end{fulllineitems}

\index{getMethodLinkSpecies\_Set\_Ids() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.getMethodLinkSpecies_Set_Ids}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{getMethodLinkSpecies\_Set\_Ids}}{\emph{speciesSet\_ids}, \emph{release}}{}
\end{fulllineitems}

\index{getSpecies\_Set\_Ids() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.getSpecies_Set_Ids}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{getSpecies\_Set\_Ids}}{\emph{genomedb\_ids}, \emph{release}}{}
\end{fulllineitems}

\index{makeTable() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.makeTable}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{makeTable}}{\emph{genomeDB\_ids}, \emph{orthoSpec2stableIds}, \emph{orthoIds}, \emph{outfile}}{}
\end{fulllineitems}

\index{q() (in module scythe2.helpers.ensembl\_ortho\_mysql)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.ensembl_ortho_mysql.q}\pysiglinewithargsret{\code{scythe2.helpers.ensembl\_ortho\_mysql.}\bfcode{q}}{\emph{s}}{}
\end{fulllineitems}

\phantomsection\label{Modules:module-scythe2.helpers.mergeSubsets}\index{scythe2.helpers.mergeSubsets (module)}\index{Orthogroup (class in scythe2.helpers.mergeSubsets)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.Orthogroup}\pysiglinewithargsret{\strong{class }\code{scythe2.helpers.mergeSubsets.}\bfcode{Orthogroup}}{\emph{list}}{}~\index{expand() (scythe2.helpers.mergeSubsets.Orthogroup method)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.Orthogroup.expand}\pysiglinewithargsret{\bfcode{expand}}{\emph{other}}{}
\end{fulllineitems}

\index{toString() (scythe2.helpers.mergeSubsets.Orthogroup method)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.Orthogroup.toString}\pysiglinewithargsret{\bfcode{toString}}{}{}
\end{fulllineitems}


\end{fulllineitems}

\index{filterGroups() (in module scythe2.helpers.mergeSubsets)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.filterGroups}\pysiglinewithargsret{\code{scythe2.helpers.mergeSubsets.}\bfcode{filterGroups}}{\emph{grp}, \emph{outfile}, \emph{numspec=None}, \emph{rename=False}}{}
\end{fulllineitems}

\index{main() (in module scythe2.helpers.mergeSubsets)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.main}\pysiglinewithargsret{\code{scythe2.helpers.mergeSubsets.}\bfcode{main}}{}{}
\end{fulllineitems}

\index{mergeSubsets() (in module scythe2.helpers.mergeSubsets)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.mergeSubsets}\pysiglinewithargsret{\code{scythe2.helpers.mergeSubsets.}\bfcode{mergeSubsets}}{\emph{grp}, \emph{outfile}, \emph{rename=False}}{}
\end{fulllineitems}

\index{usage() (in module scythe2.helpers.mergeSubsets)}

\begin{fulllineitems}
\phantomsection\label{Modules:scythe2.helpers.mergeSubsets.usage}\pysiglinewithargsret{\code{scythe2.helpers.mergeSubsets.}\bfcode{usage}}{}{}
\end{fulllineitems}



\section{algo}
\label{Modules:algo}
!todo


\chapter{Indices and tables}
\label{index:indices-and-tables}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{modindex}

\item {} 
\emph{search}

\end{itemize}

\begin{thebibliography}{NEE1970}
\bibitem[NEE1970]{NEE1970}{\phantomsection\label{Introduction:nee1970} 
Needleman, Saul B.; and Wunsch, Christian D. (1970). ``A general method applicable to the search for similarities in the amino acid sequence of two proteins''. Journal of Molecular Biology 48 (3): 443â€“53. doi:10.1016/0022-2836(70)90057-4. PMID 5420325.
}
\bibitem[EMB2000]{EMB2000}{\phantomsection\label{Introduction:emb2000} 
EMBOSS: The European Molecular Biology Open Software Suite (2000) Rice,P. Longden,I. and Bleasby, A.Trends in Genetics 16, (6) pp276--277
}
\end{thebibliography}


\renewcommand{\indexname}{Python Module Index}
\begin{theindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{s}
\item {\texttt{scythe2.convert}}, \pageref{Modules:module-scythe2.convert}
\item {\texttt{scythe2.convert.scythe\_loc\_gff}}, \pageref{Modules:module-scythe2.convert.scythe_loc_gff}
\item {\texttt{scythe2.convert.scythe\_loc\_tsv}}, \pageref{Modules:module-scythe2.convert.scythe_loc_tsv}
\item {\texttt{scythe2.gui}}, \pageref{Modules:module-scythe2.gui}
\item {\texttt{scythe2.helpers}}, \pageref{Modules:module-scythe2.helpers}
\item {\texttt{scythe2.helpers.ensembl2grp}}, \pageref{Modules:module-scythe2.helpers.ensembl2grp}
\item {\texttt{scythe2.helpers.ensembl\_ortho\_mysql}}, \pageref{Modules:module-scythe2.helpers.ensembl_ortho_mysql}
\item {\texttt{scythe2.helpers.mergeSubsets}}, \pageref{Modules:module-scythe2.helpers.mergeSubsets}
\end{theindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}
